src/astarte-client/types/dataType.ts

```diff
@@ -16,4 +16,42 @@
    limitations under the License.
 */
 
-export type AstarteDataType = number | boolean | string | number[] | boolean[] | string[];
+export type AstarteDataValue = number | boolean | string | number[] | boolean[] | string[] | null;
+
+export type AstarteDataType =
+  | 'double'
+  | 'integer'
+  | 'boolean'
+  | 'longinteger'
+  | 'string'
+  | 'binaryblob'
+  | 'datetime'
+  | 'doublearray'
+  | 'integerarray'
+  | 'booleanarray'
+  | 'longintegerarray'
+  | 'stringarray'
+  | 'binaryblobarray'
+  | 'datetimearray';
+
+export type AstarteDataTuple =
+  | { type: 'double'; value: number }
+  | { type: 'integer'; value: number }
+  | { type: 'boolean'; value: boolean }
+  | { type: 'longinteger'; value: string }
+  | { type: 'string'; value: string }
+  | { type: 'binaryblob'; value: string }
+  | { type: 'datetime'; value: string }
+  | { type: 'doublearray'; value: number[] }
+  | { type: 'integerarray'; value: number[] }
+  | { type: 'booleanarray'; value: boolean[] }
+  | { type: 'longintegerarray'; value: string[] }
+  | { type: 'stringarray'; value: string[] }
+  | { type: 'binaryblobarray'; value: string[] }
+  | { type: 'datetimearray'; value: string[] }
+  | { type: AstarteDataType; value: null };
+
+export type AstarteData = AstarteDataTuple & {
```

Comment:
While a single value (or AstarteDataTuple) belongs to a single endpoint, a single endpoint could have a list of values each with its own timestamp.
This is true for both object aggregates (may change in the future) and single path queries (`devices/{device_id}/interfaces/{interface}/{path}`).
When you query for a specific path, Astarte returns all the data for a specific endpoint, if every `AstarteData` object contains a single value, you're forced to have an array of `AstarteData` each one containing the very same endpoint.
What about adding the optional timestamp to the value itself and declaring data as such:
```
type AstarteData<AstarteDataType> = {
  endpoint: string
  value: AstarteValue<AstarteDataType> | AstarteValue<AstarteDataType>[]
}
```
Here we need to use generics to make sure the data in the array all have the same type.

Reply:
Actually I am using it as the type for linearized data

Reply:
you can linearize all trees, even the one representing properties but those don't have a timestamp.
What about returning a typed class when linearizing the tree?
Follow-up on the next comment

Reply:
Another good point. I've distinguished the two cases and got rid of type `AstarteData`. Then I reflected the two cases when linearizing trees using generics. For instance, linearizing a tree of property data will yield `AstartePropertyData[]`, while linearizing a tree with datastream data will yield `AstarteDatastreamData[]`. So either you always have a timestamp or you don't.